### PR TITLE
feat(notification): notifications to users

### PR DESCRIPTION
**Pull request by James-nyagatare**
## What does this PR do?
- Provides notifications to the Users
## Description of Task to be completed?
- Sending new trip request notification to the manager
- Sending a new notification to the requester when the trip is rejected/approved/commented by the manager
- Sending a notification to the manager when the trip is edited
- View all notification
- View all unread notification
- Mark notification as read
- Disable/enable in-app/email notification
- Mark all my notifications as read
- Document feature using swagger
## How should this be manually tested?
- Postman
- run server using npm start
- Test the real-time using the browser
## Any background context you want to provide?
- N/A
## What are the relevant pivotal tracker stories?
- [notifications](https://www.pivotaltracker.com/story/show/176153419)
- [read](https://www.pivotaltracker.com/story/show/176153453)
## Screenshots (if appropriate)
- N/A
## Questions:
- N/A

### CODE REVIEW

File: src/controllers/notificationController.js [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+  /**
+* @description this method marks one or all as read
+* @param {Object} req request
+* @param {Object} res response
+* @return {object} Oject of data or error
+* @memberof commentController
+*/
+  static async markAsRead(req, res) {
+    try {
+      const param = req.query.id ? { id: req.query.id } : { receiverId: req.user.id };
```
Comment (harerakalex) [inline src/controllers/notificationController.js:38]: Can you shed more light on this logic? @James-nyagatare

File: src/controllers/notificationController.js [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+      const data = await notificationService.markAsRead({ ...param, isRead: false });
+      let message = data > 1 ? 'Notifications' : 'Notification';
```
Comment (harerakalex) [inline src/controllers/notificationController.js:40]: Good idea.

File: src/controllers/tripController.js [hunk 3]
```diff
@@ -11,6 +11,7 @@ export default class TripController {
    * Creates a trip
    * @param {Object} req provides the requests from body to controllers
    * @param {Object} res provides responses to the users
+   * @param {Object} next provides responses to the users
```
Comment (harerakalex) [inline src/controllers/tripController.js]: Why did you add this comment?

File: src/routes/notificationRoute.js [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+ *       - application/json
+ *     responses:
+ *       200:
+ *             description: notifications succcessfuly found.
+ *       401:
+ *             description: unauthorized.
+ *       500:
+ *             description: server error.
+ * */
+router.get('/', Auth.adminAuth, Notifications.getNotifications);
```
Comment (harerakalex) [inline src/routes/notificationRoute.js]: Why did you use adminAuth? is an only admin who can see the notifications?

File: src/services/notificationServices.js [hunk 5]
```diff
@@ -0,0 +1,64 @@
...
+
+  /**
+* @description this method validate user inputs
+* @param {object} param contains valid request attributes
+* @returns {object} returns schema object
+* @memberof TripService
+*/
+  static async markAsRead(param) {
+    try {
+      const deleted = await Notification.update(
```
Comment (harerakalex) [inline src/services/notificationServices.js]: Kindly change the name of this variable.

**Review by harerakalex — CHANGES_REQUESTED** [review]
Well done @James-nyagatare, I have seen your implementation, looks good to me. KIndly consider the below comments and respond to those that need your response and implement those that need to be changed.

File: src/controllers/notificationController.js [hunk 6]
```diff
@@ -0,0 +1,50 @@
+import Response from '../helpers/sendResponse';
+import code from '../helpers/statusCode';
+import notificationService from '../services/notificationServices';
+
+/** Class representing trip controller */
+class Notifications {
```
Comment (Mucyo-Aime-Christian) [inline src/controllers/notificationController.js]: please! use valid comments

File: src/controllers/notificationController.js [hunk 7]
```diff
@@ -0,0 +1,50 @@
...
+  /**
+* @description this method gets all notifications
+* @param {Object} req request
+* @param {Object} res response
+* @return {object} Oject of data or error
+* @memberof commentController
+*/
```
Comment (Mucyo-Aime-Christian) [inline src/controllers/notificationController.js:13]: same as here

File: src/controllers/notificationController.js (hunk 1 above)
Comment (James-nyagatare) [reply src/controllers/notificationController.js:38]: The reason I used this way it's because I am using this controller for both mark all as read and mark single as read with mark all as read it will be checking the receiverId in the notifications table while with mark one as Read it will be checking the id of the single notification in Notification model and that req.query.id, I have used because in the route I wrote localhost:5000/api/v1/notification/read?id=2 that id after a question mark is what we receive as req.query in the controller.

File: src/controllers/tripController.js (hunk 3 above)
Comment (James-nyagatare) [reply src/controllers/tripController.js]: I am sorry I was trying to add create notification on a trip created using a middleware following this controller instead of using hooks and I forgot to remove when I implemented this feature with hooks.

File: src/routes/notificationRoute.js (hunk 4 above)
Comment (James-nyagatare) [reply src/routes/notificationRoute.js]: no It's because we named badly the variable but this adminAuth is the authentication of all the users I think I am going Create a bug fix story to rename this valiable as it touches different routes.

Comment (harerakalex) [reply src/routes/notificationRoute.js]: Better name it right. @James-nyagatare

**Review by harerakalex — APPROVED** [review]

**Review by Mucyo-Aime-Christian — APPROVED** [review]
LGTM.
